# Восстановление золота из руды

Необходимо подготовить прототип модели машинного обучения для компании «Цифры», которая разрабатывает решения для эффективной работы промышленных предприятий.
Модель должна предсказать коэффициент восстановления золота из золотосодержащей руды. У нас есть данные с параметрами добычи и очистки.
Модель поможет оптимизировать производство, чтобы не запускать предприятие с убыточными характеристиками.

В процессе выполнения проекта необходимо:

- Подготовить данные;
- Провести исследовательский анализ данных;
- Построить и обучить модель.

***Технологический процесс***

Когда добытая руда проходит первичную обработку, получается дроблёная смесь. Её отправляют на флотацию (обогащение) и двухэтапную очистку.


**Стадии:**

1. Флотация

Во флотационную установку подаётся смесь золотосодержащей руды. После обогащения получается черновой концентрат и «отвальные хвосты», то есть остатки продукта с низкой концентрацией ценных металлов.
На стабильность этого процесса влияет непостоянное и неоптимальное физико-химическое состояние флотационной пульпы (смеси твёрдых частиц и жидкости).

2. Очистка

Черновой концентрат проходит две очистки. На выходе получается финальный концентрат и новые отвальные хвосты.


***Описание данных***

**Технологический процесс**

* `Rougher feed` — исходное сырье
* `Rougher additions` (или `reagent additions`) — флотационные реагенты: Xanthate, Sulphate, Depressant
    * `Xanthat **` — ксантогенат (промотер, или активатор флотации);
    * `Sulphate` — сульфат (на данном производстве сульфид натрия);
    * `Depressant` — депрессант (силикат натрия).
* `Rougher process` (англ. «грубый процесс») — флотация
* `Rougher tails` — отвальные хвосты
* `Float banks` — флотационная установка
* `Cleaner process` — очистка
* `Rougher Au` — черновой концентрат золота
* `Final Au` — финальный концентрат золота

**Параметры этапов**

* `air amount` — объём воздуха
* `fluid levels` — уровень жидкости
* `feed size` — размер гранул сырья
* `feed rate` — скорость подачи


Нужно спрогнозировать сразу две величины:

1. эффективность обогащения чернового концентрата `rougher.output.recovery`;
2. эффективность обогащения финального концентрата `final.output.recovery`.


Данные находятся в трёх файлах:

* `gold_recovery_train_new_8ds.csv` — обучающая выборка;
* `gold_recovery_test_new_8ds.csv` — тестовая выборка;
* `gold_recovery_full_new_8ds.csv` — исходные данные.

Данные индексируются датой и временем получения информации (признак `date`). Соседние по времени параметры часто похожи.

Некоторые параметры недоступны, потому что замеряются и/или рассчитываются значительно позже. Из-за этого в тестовой выборке отсутствуют некоторые признаки, которые могут быть в обучающей. Также в тестовом наборе нет целевых признаков.

Исходный датасет содержит обучающую и тестовую выборки со всеми признаками.

Данные сырые. Необхоима их проверка и предобработка.

## Загрузка и подготовка данных

In [1]:
from IPython.core.display import display, HTML
display(HTML("<style>.container {width:90% !important; }</style>"))

***Выгрузка библиотек***

In [2]:
import pandas as pd
pd.set_option('display.max_columns', None)

***Константы***

In [3]:
RANDOM_SEED = 0

In [4]:
# Based on: https://stackoverflow.com/a/17303428 (для красивого оформления текста)
class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

***Функции***

In [5]:
#Рандоманая выгрузка строк
show_sample = lambda x: display(x.sample(5, random_state=RANDOM_SEED).reset_index(drop=True))

In [6]:
def com_info(data):
    '''
    Функция для изучения общей информации: input: data
                                           output: - вывод рандомных строк,
                                                   - типов данных,
                                                   - % пропусков
                                                   - статистики
                                                   - наличие дубликатов
                                                   - размерность данных
                                                   
    '''
    name = [x for x in globals() if globals()[x] is data][0]
    print(color.UNDERLINE + 'Общая информация по данным '+ name + color.END)
    show_sample(data)
    print(data.info(memory_usage=True))
    display(pd.DataFrame(round((data.isna().mean()), 2)).style.background_gradient('coolwarm')\
                                                                .set_caption('% пропусков').format({0:'{:.2%}'.format}))
    display(data.describe().T.style.set_caption('Статистика'))
    print('Количество дублей: ', data.duplicated().sum())
    print('Количество строк: ', data.shape[0], ', количество столбцов: ', data.shape[1])
    print('******************************************************************')
    print('')

In [7]:
def downcast_func(data):
    name = [x for x in globals() if globals()[x] is data][0]
    print(color.UNDERLINE + 'Информация по типам данных '+ name + color.END)
    float_features = data.select_dtypes(include='float').columns
    data[float_features] = data[float_features].apply(pd.to_numeric, downcast='float')
    print(data.info(memory_usage=True))

***Загрука данных***

In [8]:
df_train = pd.read_csv('../datasets/gold_recovery_train_new_8ds.csv')
df_test = pd.read_csv('../datasets/gold_recovery_test_new_8ds.csv')
df_full = pd.read_csv('../datasets/gold_recovery_full_new_8ds.csv')

## Изучение общей информации

In [9]:
df_list = [df_train, df_test, df_full]
for i in df_list:
    com_info(i)

Общая информация по данным df_train


date  final.output.concentrate_ag  \
0  2018-06-08 12:59:59                     4.068389   
1  2016-06-24 08:59:59                     4.672297   
2  2017-03-12 03:59:59                     4.999209   
3  2016-06-08 04:59:59                     3.879375   
4  2016-08-01 14:59:59                     4.403698   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                    11.448127                      5.983031   
1                    10.013952                      8.417600   
2                     8.771206                      8.396900   
3                     9.816033                      7.747834   
4                    10.823148                      9.911247   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    45.728369              63.185926              7.436988   
1                    45.142223              70.512351              8.217110   
2                    48.307998              66.501485              9.825832   
3                    47.631350              49.304890              8.156908   
4                    43.927370              69.502636              9.344400   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  \
0              2.577877              10.211517              2.042844   
1              1.566327              11.871072              2.433252   
2              3.145526               7.156999              3.558258   
3              2.203170               9.323327              3.720784   
4              2.947228              12.024562              3.293919   

   primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
0                      99.732267                          7.252426   
1                     107.117324                          5.006355   
2                      59.973098                          3.537027   
3                      84.321747                          8.610879   
4                     178.112944                         10.443948   

   primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
0                            6.825                        0.702993   
1                            7.690                        1.024347   
2                            7.070                        0.188916   
3                            7.080                        0.797944   
4                            7.450                        0.794329   

   primary_cleaner.output.concentrate_ag  \
0                               5.747566   
1                               8.001182   
2                               8.617695   
3                               5.900927   
4                               8.166045   

   primary_cleaner.output.concentrate_pb  \
0                              13.588954   
1                               9.587217   
2                               8.646408   
3                               4.483565   
4                              12.445559   

   primary_cleaner.output.concentrate_sol  \
0                                7.378889   
1                                8.724198   
2                                9.297762   
3                                4.621643   
4                               12.497778   

   primary_cleaner.output.concentrate_au  primary_cleaner.output.tail_ag  \
0                              34.649302                       17.597198   
1                              35.096586                       13.083296   
2                              34.263511                       16.058960   
3                              38.355818                       15.942627   
4                              34.622956                       12.861993   

   primary_cleaner.output.tail_pb  primary_cleaner.output.tail_sol  \
0                        4.702111                         6.517120   
1                        1.475464                         8.632009   
2                        3.031238                         6.169852   
3             

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14149 entries, 0 to 14148
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                14149 non-null  object 
 1   final.output.concentrate_ag                         14148 non-null  float64
 2   final.output.concentrate_pb                         14148 non-null  float64
 3   final.output.concentrate_sol                        13938 non-null  float64
 4   final.output.concentrate_au                         14149 non-null  float64
 5   final.output.recovery                               14149 non-null  float64
 6   final.output.tail_ag                                14149 non-null  float64
 7   final.output.tail_pb                                14049 non-null  float64
 8   final.output.tail_sol                               14144 non-null  float64


Количество дублей:  0
Количество строк:  14149 , количество столбцов:  87
******************************************************************

Общая информация по данным df_test


date  primary_cleaner.input.sulfate  \
0  2016-09-02 05:59:59                     126.329123   
1  2016-10-02 01:59:59                     137.116643   
2  2016-10-05 16:59:59                     136.160058   
3  2017-10-31 11:59:59                     220.706679   
4  2017-12-25 11:59:59                     212.717029   

   primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
0                          8.003034                         7.160000   
1                          3.972800                         8.600000   
2                          4.001874                         7.700000   
3                          9.139982                         7.040000   
4                         15.009052                         8.150001   

   primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
0                        0.699695                             1400.760536   
1                        0.803511                             1598.994956   
2                        0.701256                             1648.213302   
3                        2.364929                             1700.289848   
4                        0.859222                             1601.389637   

   primary_cleaner.state.floatbank8_a_level  \
0                               -500.200817   
1                               -500.061922   
2                               -500.612006   
3                               -508.068208   
4                               -399.591203   

   primary_cleaner.state.floatbank8_b_air  \
0                             1399.437375   
1                             1603.985367   
2                             1649.936185   
3                             1700.434596   
4                             1602.644217   

   primary_cleaner.state.floatbank8_b_level  \
0                               -500.679938   
1                               -499.302443   
2                               -499.849197   
3                               -489.509097   
4                               -397.938382   

   primary_cleaner.state.floatbank8_c_air  \
0                             1399.429209   
1                             1600.063771   
2                             1647.739364   
3                             1542.893209   
4                             1301.027142   

   primary_cleaner.state.floatbank8_c_level  \
0                               -499.067140   
1                               -500.351414   
2                               -500.380169   
3                               -493.814645   
4                               -402.471051   

   primary_cleaner.state.floatbank8_d_air  \
0                             1399.729437   
1                             1598.782409   
2                             1648.907802   
3                             1700.412672   
4                             1601.802270   

   primary_cleaner.state.floatbank8_d_level  rougher.input.feed_ag  \
0                               -499.640771               8.010967   
1                               -499.950738               9.962656   
2                               -500.039132              10.475888   
3                               -505.603329              12.644495   
4                               -399.593994              12.840340   

   rougher.input.feed_pb  rougher.input.feed_rate  rougher.input.feed_size  \
0               4.207135               500.255644                70.952990   
1               3.759965               399.363090                43.614246   
2               3.075634               457.442429                43.596812   
3               3.864685               632.129723                57.308352   
4               5.656224               517.436793                45.201984   

   rougher.input.feed_sol  rougher.input.feed_au  \
0               42.962929               8.287757   
1               31.639175               9.747144   
2               35.593390               9.002293   
3               43.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        5290 non-null   object 
 1   primary_cleaner.input.sulfate               5286 non-null   float64
 2   primary_cleaner.input.depressant            5285 non-null   float64
 3   primary_cleaner.input.feed_size             5290 non-null   float64
 4   primary_cleaner.input.xanthate              5286 non-null   float64
 5   primary_cleaner.state.floatbank8_a_air      5290 non-null   float64
 6   primary_cleaner.state.floatbank8_a_level    5290 non-null   float64
 7   primary_cleaner.state.floatbank8_b_air      5290 non-null   float64
 8   primary_cleaner.state.floatbank8_b_level    5290 non-null   float64
 9   primary_cleaner.state.floatbank8_c_air      5290 non-null   float64
 10  primary_clea

Количество дублей:  0
Количество строк:  5290 , количество столбцов:  53
******************************************************************

Общая информация по данным df_full


date  final.output.concentrate_ag  \
0  2017-02-08 14:59:59                     5.295988   
1  2016-10-29 13:59:59                     3.020476   
2  2017-08-10 06:59:59                     3.455687   
3  2018-03-20 17:59:59                     5.360341   
4  2017-02-26 19:59:59                     4.645203   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     8.195301                     13.379583   
1                     9.071122                     16.816693   
2                    12.662607                      6.890618   
3                    12.297913                      6.589168   
4                    10.322894                     14.283728   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    47.758663              65.280176              9.925318   
1                    49.797229              63.724514              8.165338   
2                    45.799572              62.002449              7.103502   
3                    42.586163              56.365844              8.325812   
4                    46.100397              74.086821             12.162428   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  \
0              3.387088               7.605304              3.382011   
1              3.167601              10.705567              3.591239   
2              4.485478              11.579938              3.446024   
3              1.211108               8.217830              2.221499   
4              4.429356               9.392597              3.720634   

   primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
0                      88.025736                          8.082156   
1                     149.481074                          9.316742   
2                     196.139659                          6.021495   
3                      90.885759                          9.055972   
4                     161.195780                          9.990262   

   primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
0                         6.780000                        1.349388   
1                         7.260000                        0.804833   
2                         6.900000                        0.484016   
3                         6.626667                        1.499397   
4                         7.460000                        0.799182   

   primary_cleaner.output.concentrate_ag  \
0                               8.509287   
1                               5.416631   
2                               6.686468   
3                               7.118439   
4                               9.005559   

   primary_cleaner.output.concentrate_pb  \
0                               7.432810   
1                              10.640171   
2                              14.669333   
3                               9.560626   
4                              10.509431   

   primary_cleaner.output.concentrate_sol  \
0                                6.704215   
1                                8.197248   
2                                7.538537   
3                                7.932908   
4                                8.921297   

   primary_cleaner.output.concentrate_au  primary_cleaner.output.tail_ag  \
0                              33.850846                       17.925953   
1                              40.370873                       13.716884   
2                              33.176457                       13.306054   
3                              36.147372                       19.856436   
4                              31.542245                       17.475796   

   primary_cleaner.output.tail_pb  primary_cleaner.output.tail_sol  \
0                        3.713285                         5.571651   
1                        3.021272                         9.270586   
2                        2.958052                         8.453464   
3             

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19439 entries, 0 to 19438
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                19439 non-null  object 
 1   final.output.concentrate_ag                         19438 non-null  float64
 2   final.output.concentrate_pb                         19438 non-null  float64
 3   final.output.concentrate_sol                        19228 non-null  float64
 4   final.output.concentrate_au                         19439 non-null  float64
 5   final.output.recovery                               19439 non-null  float64
 6   final.output.tail_ag                                19438 non-null  float64
 7   final.output.tail_pb                                19338 non-null  float64
 8   final.output.tail_sol                               19433 non-null  float64


Количество дублей:  0
Количество строк:  19439 , количество столбцов:  87
******************************************************************



#### Выводы

***df_train***

- Типы данных не оптимизированы, необходимо понизить разрядность признаков
- В данных есть пропуски, больше всего в признаке `secondary_cleaner.output.tail_sol` (11%). Пропусков в целевых признаках нет
- Некоторые признаки отрицательные - например, `primary_cleaner.state.floatbank8_b_level`, `primary_cleaner.state.floatbank8_c_level`, `rougher.state.floatbank10_a_level` и т.д.
- Полных дублей в данных нет
- Количество строк:  14149 , количество столбцов:  87



***df_test***

- Типы данных не оптимизированы, необходимо понизить разрядность признаков
- В данных небольшое количество пропусков по некоторым признакам - например `rougher.input.feed_sol ` , `rougher.input.floatbank11_sulfate` (меньше 1%)
- Некоторые признаки отрицательные - например, `primary_cleaner.state.floatbank8_a_level`, `rougher.state.floatbank10_b_level` и т.д.
- Полных дублей в данных нет.
- Количество строк:  5290 , количество столбцов:  53

***df_full***

- Типы данных не оптимизированы, необходимо понизить разрядность признаков
- В данных есть пропуски, больше всего в признаке  `secondary_cleaner.output.tail_sol `  (9%). Пропусков в целевых признаках нет
- Некоторые признаки отрицательные - например, `primary_cleaner.state.floatbank8_a_level`, `rougher.state.floatbank10_b_level` и т.д.
- Полных дублей в данных нет
- Количество строк:  19439 , количество столбцов:  87

## Предобработка

***Оптимизация типов данных***

In [10]:
for i in df_list:
     downcast_func(i)

Информация по типам данных df_train
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14149 entries, 0 to 14148
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                14149 non-null  object 
 1   final.output.concentrate_ag                         14148 non-null  float32
 2   final.output.concentrate_pb                         14148 non-null  float32
 3   final.output.concentrate_sol                        13938 non-null  float32
 4   final.output.concentrate_au                         14149 non-null  float32
 5   final.output.recovery                               14149 non-null  float32
 6   final.output.tail_ag                                14149 non-null  float32
 7   final.output.tail_pb                                14049 non-null  float32
 8   final.output.tail_sol                   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19439 entries, 0 to 19438
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                19439 non-null  object 
 1   final.output.concentrate_ag                         19438 non-null  float32
 2   final.output.concentrate_pb                         19438 non-null  float32
 3   final.output.concentrate_sol                        19228 non-null  float32
 4   final.output.concentrate_au                         19439 non-null  float32
 5   final.output.recovery                               19439 non-null  float32
 6   final.output.tail_ag                                19438 non-null  float32
 7   final.output.tail_pb                                19338 non-null  float32
 8   final.output.tail_sol                               19433 non-null  float32
